File: app/src/main/java/com/example/java/android1/java_android_notes/ui/ListOfNotesFragment.java
package com.example.java.android1.java_android_notes.ui;

import android.annotation.SuppressLint;
import android.content.Context;
import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.Menu;
import android.view.MenuInflater;
import android.view.MenuItem;
import android.view.View;
import android.view.ViewGroup;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.appcompat.widget.SearchView;
import androidx.fragment.app.Fragment;
import androidx.recyclerview.widget.DefaultItemAnimator;
import androidx.recyclerview.widget.DividerItemDecoration;
import androidx.recyclerview.widget.GridLayoutManager;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import com.example.java.android1.java_android_notes.MainActivity;
import com.example.java.android1.java_android_notes.R;
import com.example.java.android1.java_android_notes.Settings;
import com.example.java.android1.java_android_notes.data.DataNote;
import com.example.java.android1.java_android_notes.data.DataNoteSource;
import com.example.java.android1.java_android_notes.data.DataNoteSourceFirebase;
import com.example.java.android1.java_android_notes.dialogs.BottomSheetDialog;
import com.example.java.android1.java_android_notes.dialogs.DialogAddFragment;
import com.example.java.android1.java_android_notes.service.Navigation;
import com.example.java.android1.java_android_notes.service.NotesAdapter;
import com.google.android.material.floatingactionbutton.FloatingActionButton;

import java.util.LinkedList;
import java.util.Locale;

public class ListOfNotesFragment extends Fragment {

    public static final String KEY_NOTE_POSITION = "NotesFragment.notesPosition";
    private static final String KEY_BOTTOM_SHEET_DIALOG_NOTE = "ListOfNotesFragment.BottomSheetDialog.AboutNote";
    private static final String KEY_DIALOG_ADD_NOTE = "ListOfNotesFragment.Dialog.AddNote";

    private NotesAdapter mNotesAdapter;
    private DataNoteSource mDataNoteSource;
    private int mItemIndex = -1;
    private RecyclerView mRecyclerView;
    private Navigation mNavigation;

    @SuppressLint("NotifyDataSetChanged")
    private final DataNoteSource.DataNoteSourceListener mChangeListener = new DataNoteSource.DataNoteSourceListener() {
        @Override
        public void onItemAdded(int index) {
            if (mNotesAdapter != null) {
                mNotesAdapter.notifyItemInserted(index);
                mRecyclerView.scrollToPosition(index);
            }
        }

        @Override
        public void onItemRemoved(int index) {
            if (mNotesAdapter != null) {
                mNotesAdapter.notifyItemRemoved(index);
                mItemIndex = index - 1;
            }
        }

        @Override
        public void onItemUpdated(int index) {
            if (mNotesAdapter != null) {
                mNotesAdapter.notifyItemChanged(index);
            }
        }

        @Override
        public void onDataSetChanged() {
            if (mNotesAdapter != null) {
                mNotesAdapter.notifyDataSetChanged();
            }
        }
    };

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
    }

    @Override
    public void onSaveInstanceState(@NonNull Bundle outState) {
        super.onSaveInstanceState(outState);
        outState.putInt(KEY_NOTE_POSITION, mItemIndex);
    }

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container,
                             Bundle savedInstanceState) {
        ViewGroup viewGroup = (ViewGroup) inflater.inflate(R.layout.fragment_list_of_notes, container, false);
        setHasOptionsMenu(true);
        initNotes(viewGroup);
        mNavigation = new Navigation(requireActivity().getSupportFragmentManager(),
                (MainActivity) requireActivity());
        return viewGroup;
    }

    public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);
        if (savedInstanceState != null) {
            mItemIndex = savedInstanceState.getInt(KEY_NOTE_POSITION, -1);
        }
        addNote(view);
    }

    private void initNotes(ViewGroup view) {
        mRecyclerView = view.findViewById(R.id.list_of_notes_container);
        mRecyclerView.setHasFixedSize(true);

        mDataNoteSource = DataNoteSourceFirebase.getInstance();
        mNotesAdapter = new NotesAdapter(this, mDataNoteSource);
        setDecoration();
        setAnimation(1000);
        mRecyclerView.setAdapter(mNotesAdapter);

        mNotesAdapter.setOnItemClickListener((click, position) -> {
            mItemIndex = position;
            NoteDescriptionFragment fragment = NoteDescriptionFragment.newInstance(mItemIndex);
            mNavigation.addFragment(fragment, true);
        });

        mNotesAdapter.setOnItemLongClickListener(position -> new BottomSheetDialog(position).
                show(getParentFragmentManager(), KEY_BOTTOM_SHEET_DIALOG_NOTE));

        setLayoutManager();
        mDataNoteSource.addChangesListener(mChangeListener);
    }

    @Override
    public void onCreateOptionsMenu(@NonNull Menu menu, @NonNull MenuInflater inflater) {
        super.onCreateOptionsMenu(menu, inflater);
        MenuInflater menuInflater = requireActivity().getMenuInflater();
        menuInflater.inflate(R.menu.main_menu, menu);
        searchNote(menu);
    }

    @Override
    public boolean onOptionsItemSelected(@NonNull MenuItem item) {
        if (item.getItemId() == R.id.action_sort) {
            mDataNoteSource.sortListByDate();
        } else {
            return super.onOptionsItemSelected(item);
        }
        return true;
    }

    private void searchNote(Menu menu) {
        MenuItem search = menu.findItem(R.id.action_search);
        SearchView searchView = (SearchView) search.getActionView();
        searchView.setOnQueryTextListener(new SearchView.OnQueryTextListener() {
            @Override
            public boolean onQueryTextSubmit(String query) {
                return true;
            }

            @Override
            public boolean onQueryTextChange(String newText) {
                // ???????? ?????????? ???????????? ?????????? ?????????? ?????????????? ??????????????.
                searchFilter(newText);
                return true;
            }
        });
    }

    private void searchFilter(String text) {
        LinkedList<DataNote> filterList = new LinkedList<>();
        for (DataNote note : mDataNoteSource.getDataNote()) {
            if (note.getNoteName().toLowerCase(Locale.ROOT).contains(text.toLowerCase(Locale.ROOT))) {
                filterList.add(note);
            }
        }
        mDataNoteSource.filterList(filterList);
        if (text.trim().equals("")) {
            mDataNoteSource.recreateList();
        }
    }

    private void addNote(View view) {
        FloatingActionButton actionButton = view.findViewById(R.id.create_note);
        actionButton.setOnClickListener((click) -> new DialogAddFragment().show(requireActivity().
                getSupportFragmentManager(), KEY_DIALOG_ADD_NOTE));
    }

    public void setLastSelectedPosition(int lastSelectedPosition) {
        mItemIndex = lastSelectedPosition;
    }

    @Override
    public void onDestroyView() {
        super.onDestroyView();
        mDataNoteSource.removeChangesListener(mChangeListener);
    }

    private void setLayoutManager() {
        boolean isGridLayoutView = requireActivity().getSharedPreferences(Settings.SHARED_PREFERENCE_NAME,
                Context.MODE_PRIVATE).getInt(Settings.KEY_LAYOUT_VIEW, 1) == Settings.GRID_LAYOUT_VIEW;

        if (isGridLayoutView) {
            GridLayoutManager layoutManager = new GridLayoutManager(requireActivity(), 2);
            mRecyclerView.setLayoutManager(layoutManager);
        } else {
            LinearLayoutManager layoutManager = new LinearLayoutManager(requireActivity());
            mRecyclerView.setLayoutManager(layoutManager);
        }

    }

    @SuppressLint("UseCompatLoadingForDrawables")
    private void setDecoration() {
        DividerItemDecoration decoration = new DividerItemDecoration(requireActivity(), LinearLayoutManager.VERTICAL);
        decoration.setDrawable(getResources().getDrawable(R.drawable.separator));
        mRecyclerView.addItemDecoration(decoration);
    }

    private void setAnimation(long duration) {
        DefaultItemAnimator animator = new DefaultItemAnimator();
        animator.setAddDuration(duration);
        animator.setRemoveDuration(duration);
        mRecyclerView.setItemAnimator(animator);
    }

}

File: app/src/main/java/com/example/java/android1/java_android_notes/ui/ListOfNotesFragment.kt
package com.example.java.android1.java_android_notes.ui

import android.annotation.SuppressLint
import android.content.Context
import android.os.Bundle
import android.view.*
import androidx.appcompat.widget.SearchView
import androidx.fragment.app.Fragment
import androidx.recyclerview.widget.*
import com.example.java.android1.java_android_notes.MainActivity
import com.example.java.android1.java_android_notes.R
import com.example.java.android1.java_android_notes.Settings
import com.example.java.android1.java_android_notes.data.DataNote
import com.example.java.android1.java_android_notes.data.DataNoteSource
import com.example.java.android1.java_android_notes.data.DataNoteSource.DataNoteSourceListener
import com.example.java.android1.java_android_notes.data.DataNoteSourceFirebase
import com.example.java.android1.java_android_notes.databinding.FragmentListOfNotesBinding
import com.example.java.android1.java_android_notes.dialogs.BottomSheetDialog
import com.example.java.android1.java_android_notes.dialogs.DialogAddFragment
import com.example.java.android1.java_android_notes.service.Navigation
import com.example.java.android1.java_android_notes.service.NotesAdapter
import java.util.*

class ListOfNotesFragment : Fragment() {
    private var dataNoteSource: DataNoteSource? = null
    private var itemIndex = -1
    private var navigation: Navigation? = null
    private val notesAdapter: NotesAdapter by lazy {
        NotesAdapter(
            notesFragment = this,
            dataNoteSource = dataNoteSource,
            onItemClickListener = { position ->
                itemIndex = position
                val fragment: NoteDescriptionFragment =
                    NoteDescriptionFragment.newInstance(itemIndex)
                navigation?.addFragment(fragment, true)
            },
            onItemLongClickListener = { position ->
                BottomSheetDialog(position).show(
                    parentFragmentManager, KEY_BOTTOM_SHEET_DIALOG_NOTE
                )
            })
    }

    private var _binding: FragmentListOfNotesBinding? = null
    private val binding get() = _binding!!

    @SuppressLint("NotifyDataSetChanged")
    private val changeListener: DataNoteSourceListener = object : DataNoteSourceListener {
        override fun onItemAdded(index: Int) {
            notesAdapter.notifyItemInserted(index)
            binding.listOfNotesContainer.scrollToPosition(index)
        }

        override fun onItemRemoved(index: Int) {
            notesAdapter.notifyItemRemoved(index)
            itemIndex = index - 1
        }

        override fun onItemUpdated(index: Int) {
            notesAdapter.notifyItemChanged(index)
        }

        override fun onDataSetChanged() {
            notesAdapter.notifyDataSetChanged()
        }
    }

    override fun onSaveInstanceState(outState: Bundle) {
        super.onSaveInstanceState(outState)
        outState.putInt(KEY_NOTE_POSITION, itemIndex)
    }

    override fun onCreateView(
        inflater: LayoutInflater, container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View {
        _binding = FragmentListOfNotesBinding.inflate(inflater)
        setHasOptionsMenu(true)
        initNotes()
        navigation = Navigation(
            requireActivity().supportFragmentManager,
            (requireActivity() as MainActivity)
        )
        if (savedInstanceState != null) {
            itemIndex = savedInstanceState.getInt(KEY_NOTE_POSITION, -1)
        }
        addNote()
        return binding.root
    }

    private fun initNotes() {
        dataNoteSource = DataNoteSourceFirebase.instance
        binding.listOfNotesContainer.apply {
            setHasFixedSize(true)
            adapter = notesAdapter
        }
        setDecoration()
        setAnimation(1000)
        setLayoutManager()
        dataNoteSource?.addChangesListener(changeListener)
    }

    override fun onCreateOptionsMenu(menu: Menu, inflater: MenuInflater) {
        super.onCreateOptionsMenu(menu, inflater)
        val menuInflater = requireActivity().menuInflater
        menuInflater.inflate(R.menu.main_menu, menu)
        searchNote(menu)
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        if (item.itemId == R.id.action_sort) {
            dataNoteSource?.sortListByDate()
        } else {
            return super.onOptionsItemSelected(item)
        }
        return true
    }

    private fun searchNote(menu: Menu) {
        val search = menu.findItem(R.id.action_search)
        val searchView = search.actionView as SearchView
        searchView.setOnQueryTextListener(object : SearchView.OnQueryTextListener {
            override fun onQueryTextSubmit(query: String): Boolean {
                return true
            }

            override fun onQueryTextChange(newText: String): Boolean {
                searchFilter(newText)
                return true
            }
        })
    }

    private fun searchFilter(text: String) {
        val filterList = LinkedList<DataNote>()
        dataNoteSource?.dataNote?.forEach { dataNote ->
            if (dataNote.noteName?.lowercase()?.contains(text.lowercase()) == true) {
                filterList.add(dataNote)
            }
        }
        dataNoteSource?.filterList(filterList)
        if (text.trim { it <= ' ' } == "") {
            dataNoteSource?.recreateList()
        }
    }

    private fun addNote() {
        binding.createNote.setOnClickListener {
            DialogAddFragment().show(
                requireActivity().supportFragmentManager,
                KEY_DIALOG_ADD_NOTE
            )
        }
    }

    fun setLastSelectedPosition(lastSelectedPosition: Int) {
        itemIndex = lastSelectedPosition
    }

    override fun onDestroyView() {
        super.onDestroyView()
        dataNoteSource?.removeChangesListener(changeListener)
    }

    private fun setLayoutManager() {
        val isGridLayoutView = requireActivity().getSharedPreferences(
            Settings.SHARED_PREFERENCE_NAME,
            Context.MODE_PRIVATE
        ).getInt(Settings.KEY_LAYOUT_VIEW, 1) == Settings.GRID_LAYOUT_VIEW
        if (isGridLayoutView) {
            val layoutManager = GridLayoutManager(requireActivity(), 2)
            binding.listOfNotesContainer.layoutManager = layoutManager
        } else {
            val layoutManager = LinearLayoutManager(requireActivity())
            binding.listOfNotesContainer.layoutManager = layoutManager
        }
    }

    @SuppressLint("UseCompatLoadingForDrawables")
    private fun setDecoration() {
        val decoration = DividerItemDecoration(requireActivity(), LinearLayoutManager.VERTICAL)
        decoration.setDrawable(resources.getDrawable(R.drawable.separator))
        binding.listOfNotesContainer.addItemDecoration(decoration)
    }

    private fun setAnimation(duration: Long) {
        val animator = DefaultItemAnimator()
        animator.addDuration = duration
        animator.removeDuration = duration
        binding.listOfNotesContainer.itemAnimator = animator
    }

    companion object {
        const val KEY_NOTE_POSITION = "NotesFragment.notesPosition"
        private const val KEY_BOTTOM_SHEET_DIALOG_NOTE =
            "ListOfNotesFragment.BottomSheetDialog.AboutNote"
        private const val KEY_DIALOG_ADD_NOTE = "ListOfNotesFragment.Dialog.AddNote"
    }
}